### PR TITLE
Opt physics experience and opt physics unit test

### PR DESCRIPTION
**Pull request by luzhuang**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## Summary by CodeRabbit

- **New Features**
	- Enhanced property management in the `CharacterController`, `Collider`, and `DynamicCollider` classes.
	- New methods for retrieving physical properties in `IDynamicCollider` and `PhysXDynamicCollider`.
	- Added `setContactOffset` method in `PhysXCharacterController`.
	- Introduced new internal properties and methods for managing collider shapes in `PhysXCollider` and `LiteCollider`.
	- Updated handling of scaling in `PhysXBoxColliderShape` and `PhysXCapsuleColliderShape` to utilize `_worldScale`.
	- Removed swing offset functionality from `HingeJoint` and `SpringJoint`.

- **Bug Fixes**
	- Improved error handling for shape management in various classes.

- **Documentation**
	- Updated documentation for properties and methods across multiple classes to clarify usage and constraints.

- **Tests**
	- Expanded and refined test suites for `CharacterController`, `DynamicCollider`, and various joint classes, improving coverage and error handling.
	- Added tests for cloning functionality in `ColliderShape` and verified the handling of `contactOffset`.

### CODE REVIEW

File: packages/core/src/physics/shape/BoxColliderShape.ts [hunk 1]
```diff
@@ -32,9 +32,8 @@ export class BoxColliderShape extends ColliderShape {
       this._material._nativeMaterial
     );
 
-    this._setSize = this._setSize.bind(this);
     //@ts-ignore
-    this._size._onValueChanged = this._setSize;
+    this._size._onValueChanged = this._setSize.bind(this);
```
[the remark this answers is not in the document]
Comment (luzhuang) [reply packages/core/src/physics/shape/BoxColliderShape.ts:36]: ignore

File: packages/core/src/physics/CharacterController.ts [hunk 2]
```diff
@@ -14,9 +14,24 @@ export class CharacterController extends Collider {
...
+   */
+  get contactOffset(): number {
+    return this._contactOffset;
+  }
+
+  set contactOffset(value: number) {
+    if (this._contactOffset !== value) {
+      this._contactOffset = value;
+      (<ICharacterController>this._nativeCollider).setContactOffset(value);
+    }
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:30]: `contactOffset` should large than 0?

File: packages/core/src/physics/CharacterController.ts [hunk 3]
```diff
@@ -57,7 +72,7 @@ export class CharacterController extends Collider {
   }
 
   /**
-   * The slope limit for the controller.
+   * The slope limit for the controller, the value is the cosine value of the maximum slope angle, the default value is 0.707, which is the cosine value of 45 degrees.
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:75]: Use `@defaultValue` 0.707 specification

File: packages/core/src/physics/Collider.ts [hunk 4]
```diff
@@ -130,17 +124,20 @@ export class Collider extends Component implements ICustomClone {
...
    * @internal
    */
   _cloneTo(target: Collider): void {
-    const shapes = target._shapes;
-    for (let i = 0, n = shapes.length; i < n; i++) {
-      target._addPhysicsShape(shapes[i]);
+    target._phasedActiveInScene && target._syncBackends();
+  }
+
+  protected _syncBackends(): void {
```
Comment (GuoLei1990) [inline packages/core/src/physics/Collider.ts:138]: `_syncNative`? Suggested unified concept

Comment (luzhuang) [reply packages/core/src/physics/Collider.ts:138]: done

File: packages/core/src/physics/CharacterController.ts (hunk 3 above)
Comment (luzhuang) [reply packages/core/src/physics/CharacterController.ts:75]: done

File: packages/core/src/physics/CharacterController.ts (hunk 2 above)
Comment (GuoLei1990) [reply packages/core/src/physics/CharacterController.ts:30]: Use Math.max

Comment (luzhuang) [reply packages/core/src/physics/CharacterController.ts:31]: done

File: packages/core/src/physics/CharacterController.ts [hunk 5]
```diff
@@ -10,13 +10,30 @@ import { ColliderShape } from "./shape";
...
+  get contactOffset(): number {
+    return this._contactOffset;
+  }
+
+  set contactOffset(value: number) {
+    value = Math.max(0, value);
+    if (this._contactOffset !== value) {
+      this._contactOffset = value;
+      (<ICharacterController>this._nativeCollider).setContactOffset(value);
+    }
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:31]: Try best use shape's `contactOffset`

File: packages/core/src/physics/CharacterController.ts [hunk 6]
```diff
@@ -10,13 +10,30 @@ import { ColliderShape } from "./shape";
...
+  }
 
   /**
    * The step offset for the controller.
+   * @remarks Character can overcome obstacle less than the height(stepOffset + contractOffset(0.1)).
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:36]: Maybe should modify when opt contractOffset

File: packages/core/src/physics/CharacterController.ts [hunk 7]
```diff
@@ -10,13 +10,30 @@ import { ColliderShape } from "./shape";
  * The character controllers.
  */
 export class CharacterController extends Collider {
-  private _stepOffset: number = 0.5;
+  private _stepOffset = 0.5;
   private _nonWalkableMode: ControllerNonWalkableMode = ControllerNonWalkableMode.PreventClimbing;
   private _upDirection = new Vector3(0, 1, 0);
-  private _slopeLimit: number = 0.707;
+  private _slopeLimit = 0.707;
```
[the remark this answers is not in the document]
Comment (luzhuang) [reply packages/core/src/physics/CharacterController.ts:18]: done

File: packages/core/src/physics/DynamicCollider.ts [hunk 8]
```diff
@@ -324,37 +405,58 @@ export class DynamicCollider extends Collider {
...
+    (<IDynamicCollider>this._nativeCollider).setSolverIterations(this._solverIterations);
+    (<IDynamicCollider>this._nativeCollider).setIsKinematic(this._isKinematic);
+    (<IDynamicCollider>this._nativeCollider).setConstraints(this._constraints);
+    (<IDynamicCollider>this._nativeCollider).setCollisionDetectionMode(this._collisionDetectionMode);
+  }
+
+  private _setMassAndUpdateInertia(): void {
+    (<IDynamicCollider>this._nativeCollider).setMassAndUpdateInertia(this._mass);
+    this._setCenterOfMass();
+    this._setInertiaTensor();
```
Comment (GuoLei1990) [inline packages/core/src/physics/DynamicCollider.ts:440]:
```ts
 private _setMassAndUpdateInertia(): void {
    (<IDynamicCollider>this._nativeCollider).setMassAndUpdateInertia(this._mass);
    if (!this._automaticCenterOfMass) {
      this._setCenterOfMass();
    }
    if (!this._automaticInertiaTensor) {
      this._setInertiaTensor();
    }
  }
```

File: packages/core/src/physics/DynamicCollider.ts [hunk 9]
```diff
@@ -103,31 +97,89 @@ export class DynamicCollider extends Collider {
...
   }
 
   /**
    * The center of mass relative to the transform's origin.
+   * @remarks The center of mass is automatically calculated, if you want to set it manually, please set automaticCenterOfMass to false.
    */
   get centerOfMass(): Vector3 {
+    if (this._automaticCenterOfMass) {
+      this._setMassAndUpdateInertia();
+    }
```
Comment (GuoLei1990) [inline packages/core/src/physics/DynamicCollider.ts:131]: Why call this

File: packages/core/src/physics/DynamicCollider.ts [hunk 10]
```diff
@@ -103,31 +97,89 @@ export class DynamicCollider extends Collider {
...
+  }
+
   /**
    * The diagonal inertia tensor of mass relative to the center of mass.
+   * @remarks The inertia tensor is automatically calculated, if you want to set it manually, please set automaticInertiaTensor to false.
    */
   get inertiaTensor(): Vector3 {
+    if (this._automaticInertiaTensor) {
+      this._setMassAndUpdateInertia();
+    }
```
Comment (GuoLei1990) [inline packages/core/src/physics/DynamicCollider.ts:171]: same with before

File: packages/core/src/physics/CharacterController.ts (hunk 6 above)
Comment (luzhuang) [reply packages/core/src/physics/CharacterController.ts:22]: done

File: packages/core/src/physics/DynamicCollider.ts (hunk 8 above)
Comment (luzhuang) [reply packages/core/src/physics/DynamicCollider.ts]: done

File: packages/core/src/physics/DynamicCollider.ts (hunk 9 above)
Comment (luzhuang) [reply packages/core/src/physics/DynamicCollider.ts]: done

File: packages/core/src/physics/DynamicCollider.ts (hunk 10 above)
Comment (luzhuang) [reply packages/core/src/physics/DynamicCollider.ts]: done

File: packages/core/src/physics/Collider.ts [hunk 11]
```diff
@@ -79,9 +74,7 @@ export class Collider extends Component implements ICustomClone {
     const shapes = this._shapes;
     for (let i = 0, n = shapes.length; i < n; i++) {
       const shape = shapes[i];
-      this._phasedActiveInScene && this.scene.physics._removeColliderShape(shape);
-      shape._destroy();
-      this._nativeCollider.removeShape(shape._nativeShape);
+      this._removeNativeShape(shape);
```
Comment (GuoLei1990) [inline packages/core/src/physics/Collider.ts:77]:
```ts
for (let i = 0, n = shapes.length; i < n; i++) {
      this._removeNativeShape(shapes[i]);
    }
```

File: packages/core/src/physics/DynamicCollider.ts [hunk 12]
```diff
@@ -103,14 +105,40 @@ export class DynamicCollider extends Collider {
...
+        this._setMassAndUpdateInertia();
+      } else {
+        (<IDynamicCollider>this._nativeCollider).setMass(value);
+      }
+    }
+  }
+
+  /**
+   * Whether or not to calculate the center of mass automatically, if true, the center of mass will be calculated based on the shapes added to the collider.
+   */
```
Comment (GuoLei1990) [inline packages/core/src/physics/DynamicCollider.ts:118]: I think is not only `on the shapes added to the collider`?

File: packages/core/src/physics/DynamicCollider.ts [hunk 13]
```diff
@@ -324,37 +403,76 @@ export class DynamicCollider extends Collider {
...
   }
 
+  private _setMassAndUpdateInertia(): void {
+    (<IDynamicCollider>this._nativeCollider).setMassAndUpdateInertia(this._mass);
+
+    !this._automaticCenterOfMass && (<IDynamicCollider>this._nativeCollider).setCenterOfMass(this._centerOfMass);
+    !this._automaticInertiaTensor && (<IDynamicCollider>this._nativeCollider).setInertiaTensor(this._inertiaTensor);
+  }
+
+  private _setLinearVelocity(): void {
```
Comment (GuoLei1990) [inline packages/core/src/physics/DynamicCollider.ts:451]: Add ` @ignoreClone`

File: packages/core/src/physics/PhysicsMaterial.ts [hunk 14]
```diff
@@ -6,9 +6,9 @@ import { PhysicsMaterialCombineMode } from "./enums/PhysicsMaterialCombineMode";
  * Material class to represent a set of surface properties.
  */
 export class PhysicsMaterial {
-  private _bounciness: number = 0.1;
-  private _dynamicFriction: number = 0.1;
-  private _staticFriction: number = 0.1;
+  private _bounciness: number = 0;
+  private _dynamicFriction: number = 0;
+  private _staticFriction: number = 0;
```
Comment (GuoLei1990) [inline packages/core/src/physics/PhysicsMaterial.ts:11]: Do you think 0 is the best default value?

File: packages/physics-physx/src/PhysXCharacterController.ts [hunk 15]
```diff
@@ -54,6 +56,13 @@ export class PhysXCharacterController implements ICharacterController {
     }
   }
 
+  /**
+   * {@inheritDoc ICharacterController.setContactOffset }
+   */
+  setContactOffset(offset: number): void {
+    this._pxController?.setContactOffset(offset);
+  }
```
Comment (GuoLei1990) [inline packages/physics-physx/src/PhysXCharacterController.ts:64]: Delete this

File: packages/physics-physx/src/PhysXCharacterController.ts [hunk 16]
```diff
@@ -90,6 +99,8 @@ export class PhysXCharacterController implements ICharacterController {
     this._pxManager && this._createPXController(this._pxManager, shape);
     this._shape = shape;
     shape._controllers.add(this);
+    this.setContactOffset(shape._contractOffset);
+    this._scene?._addColliderShape(shape._id);
```
Comment (GuoLei1990) [inline packages/physics-physx/src/PhysXCharacterController.ts:103]: Why only character need this

File: packages/core/src/physics/shape/ColliderShape.ts [hunk 17]
```diff
@@ -53,14 +51,14 @@ export abstract class ColliderShape implements ICustomClone {
...
-   * Contact offset for this shape.
+   * Contact offset for this shape, the value must be greater than or equal to 0.
    */
   get contactOffset(): number {
     return this._contactOffset;
   }
 
   set contactOffset(value: number) {
-    if (this._contactOffset !== value) {
+    if (this._contactOffset !== value && value >= 0) {
```
Comment (GuoLei1990) [inline packages/core/src/physics/shape/ColliderShape.ts:61]: use Math.min

File: packages/core/src/physics/shape/ColliderShape.ts [hunk 18]
```diff
@@ -130,32 +128,41 @@ export abstract class ColliderShape implements ICustomClone {
...
   _destroy() {
-    this._material._destroy();
     this._nativeShape.destroy();
+    this._nativeShape = null;
+    delete Engine._physicalObjectsMap[this._id];
+  }
+
+  protected _syncNative(): void {
+    this._setPosition();
+    this._setRotation();
```
Comment (GuoLei1990) [inline packages/core/src/physics/shape/ColliderShape.ts:153]: `this._collider?._handleShapesChanged();` will called twice

File: packages/physics-physx/src/PhysXCharacterController.ts (hunk 16 above)
Comment (luzhuang) [reply packages/physics-physx/src/PhysXCharacterController.ts:103]: sync shape contractOffset to character

File: packages/core/src/physics/Collider.ts (hunk 11 above)
Comment (luzhuang) [reply packages/core/src/physics/Collider.ts:77]: done

File: packages/physics-physx/src/PhysXCharacterController.ts (hunk 15 above)
Comment (luzhuang) [reply packages/physics-physx/src/PhysXCharacterController.ts:64]: done

File: packages/core/src/physics/shape/ColliderShape.ts (hunk 17 above)
Comment (luzhuang) [reply packages/core/src/physics/shape/ColliderShape.ts:61]: done

File: packages/physics-physx/src/PhysXCollider.ts [hunk 19]
```diff
@@ -26,13 +31,19 @@ export abstract class PhysXCollider implements ICollider {
...
   }
 
   /**
    * {@inheritDoc ICollider.removeShape }
    */
   removeShape(shape: PhysXColliderShape): void {
     this._pxActor.detachShape(shape._pxShape, true);
+    const shapeIds = this._shapeIds;
+    const index = shapeIds.indexOf(shape._id);
+    index > -1 && shapeIds.splice(index, 1);
```
Comment (GuoLei1990) [inline packages/physics-physx/src/PhysXCollider.ts:45]: Should delete `index > -1 &&`

File: packages/physics-physx/src/PhysXDynamicCollider.ts [hunk 20]
```diff
@@ -79,13 +117,42 @@ export class PhysXDynamicCollider extends PhysXCollider implements IDynamicColli
...
+  setMassAndUpdateInertia(mass: number): void {
+    this._pxActor.setMassAndUpdateInertia(mass);
+  }
+
+  /**
+   * {@inheritDoc IDynamicCollider.getMaxAngularVelocity }
+   */
+  getMaxAngularVelocity(): number {
+    return this._pxActor.getMaxAngularVelocity();
+  }
```
Comment (GuoLei1990) [inline packages/physics-physx/src/PhysXDynamicCollider.ts:140]: If never used, can delete to reduce wasm size

File: packages/physics-physx/src/PhysXPhysicsScene.ts [hunk 21]
```diff
@@ -167,7 +151,11 @@ export class PhysXPhysicsScene implements IPhysicsScene {
    * {@inheritDoc IPhysicsManager.removeCharacterController }
    */
   removeCharacterController(characterController: PhysXCharacterController): void {
+    characterController._scene = null;
     characterController._pxManager = null;
+    characterController._destroyPXController();
+    const shape = characterController._shape;
+    shape && this._removeColliderShape(shape?._id);
```
Comment (GuoLei1990) [inline packages/physics-physx/src/PhysXPhysicsScene.ts:158]: `shape?._id`->`shape._id`

File: packages/physics-physx/src/PhysXDynamicCollider.ts (hunk 20 above)
Comment (luzhuang) [reply packages/physics-physx/src/PhysXDynamicCollider.ts:140]: done

File: packages/physics-physx/src/PhysXCollider.ts (hunk 19 above)
Comment (luzhuang) [reply packages/physics-physx/src/PhysXCollider.ts]: done

File: packages/core/src/physics/DynamicCollider.ts [hunk 22]
```diff
@@ -120,10 +148,32 @@ export class DynamicCollider extends Collider {
     }
   }
 
+  /**
+   * Whether or not to calculate the inertia tensor automatically, if true, the inertia tensor will be calculated based on the shapes added to the collider.
+   */
```
[the remark this answers is not in the document]
Comment (luzhuang) [reply packages/core/src/physics/DynamicCollider.ts:152]: opted

File: packages/core/src/physics/PhysicsMaterial.ts (hunk 14 above)
Comment (luzhuang) [reply packages/core/src/physics/PhysicsMaterial.ts:11]: changed to 0.5

File: packages/core/src/physics/DynamicCollider.ts (hunk 13 above)
Comment (luzhuang) [reply packages/core/src/physics/DynamicCollider.ts:430]: done

File: packages/core/src/physics/DynamicCollider.ts (hunk 12 above)
Comment (luzhuang) [reply packages/core/src/physics/DynamicCollider.ts:117]: opted

File: packages/physics-physx/src/joint/PhysXHingeJoint.ts [hunk 23]
```diff
@@ -8,17 +8,20 @@ import { PhysXJoint } from "./PhysXJoint";
  * A joint which behaves in a similar way to a hinge or axle.
  */
 export class PhysXHingeJoint extends PhysXJoint implements IHingeJoint {
+  protected static _xAxis = new Vector3(1, 0, 0);
+
+  private _anchor = new Vector3();
```
[the remark this answers is not in the document]
Comment (luzhuang) [reply packages/physics-physx/src/joint/PhysXHingeJoint.ts:13]: done

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**